File: web/src/auth/roles.ts
export const RoleGuest: number = 0
export const RoleRegularMember: number = 1
export const RoleCertifiedMember: number = 2
export const RoleClassDeputySecretary: number = 3
export const RoleClassSecretary: number = 4
export const RoleDeputySecretary: number = 5
export const RoleSecretary: number = 6
export const RoleRoot: number = 7

export const RoleGroupGuest: number = 0
export const RoleGroupMember: number = 1
export const RoleGroupClassManager: number = 2
export const RoleGroupGlobalManager: number = 3
export const RoleGroupRoot: number = 4

export function GetRoleName(role: number): string {
    switch (role) {
        case RoleRegularMember:
            return "Đoàn viên (Chưa gia nhập)"
        case RoleCertifiedMember:
            return "Đoàn viên"
        case RoleClassDeputySecretary:
            return "Phó bí thư Chi Đoàn"
        case RoleClassSecretary:
            return "Bí thư Chi Đoàn"
        case RoleDeputySecretary:
            return "Phó bí thư Đoàn"
        case RoleSecretary:
            return "Bí thư Đoàn"
        case RoleRoot:
            return "-"
    }
    return "Khách"
}

export function GetRoleTable(): { role: number; name: string }[] {
    return [
        RoleGuest,
        RoleRegularMember,
        RoleCertifiedMember,
        RoleClassDeputySecretary,
        RoleClassSecretary,
        RoleDeputySecretary,
        RoleSecretary,
        //RoleRoot, // hidden
    ].map(v => {
        return {
            "role": v,
            "name": GetRoleName(v)
        }
    })
}

export function GetRoleGroup(role: number): number {
    switch (role) {
        case RoleGuest:
            return RoleGroupGuest
        case RoleCertifiedMember:
        case RoleRegularMember:
            return RoleGroupMember
        case RoleClassSecretary:
        case RoleClassDeputySecretary:
            return RoleGroupClassManager
        case RoleSecretary:
        case RoleDeputySecretary:
            return RoleGroupGlobalManager
        case RoleRoot:
            return RoleGroupRoot
    }
    return RoleGuest
}

export function IsMember(role: number): boolean {
    return role == RoleRegularMember || role == RoleCertifiedMember
}

export function IsManager(role: number): boolean {
    return GetRoleGroup(role) == RoleGroupClassManager || GetRoleGroup(role) == RoleGroupGlobalManager || GetRoleGroup(role) == RoleGroupRoot
}

export function IsClassRole(role: number): boolean {
    return IsMember(role) || GetRoleGroup(role) == RoleGroupClassManager
}

export function IsLoggedIn(role: number): boolean {
    return role != RoleGuest
}

export function IsCertified(role: number): boolean {
    return role != RoleGuest && role != RoleRegularMember
}

File: server/security/roles.go
package security

const RoleGuest uint8 = 0
const RoleRegularMember uint8 = 1
const RoleCertifiedMember uint8 = 2
const RoleClassDeputySecretary uint8 = 3
const RoleClassSecretary uint8 = 4
const RoleDeputySecretary uint8 = 5
const RoleSecretary uint8 = 6
const RoleRoot uint8 = 7

const RoleGroupGuest uint8 = 0
const RoleGroupMember uint8 = 1
const RoleGroupClassManager uint8 = 2
const RoleGroupGlobalManager uint8 = 3
const RoleGroupRoot uint8 = 4

func GetRoleGroup(role uint8) uint8 {
	switch role {
	case RoleGuest:
		return RoleGroupGuest
	case RoleCertifiedMember, RoleRegularMember:
		return RoleGroupMember
	case RoleClassSecretary, RoleClassDeputySecretary:
		return RoleGroupClassManager
	case RoleSecretary, RoleDeputySecretary:
		return RoleGroupGlobalManager
	case RoleRoot:
		return RoleGroupRoot
	}
	return RoleGuest
}

func IsMember(role uint8) bool {
	return role == RoleRegularMember || role == RoleCertifiedMember
}

func IsManager(role uint8) bool {
	return GetRoleGroup(role) == RoleGroupClassManager || GetRoleGroup(role) == RoleGroupGlobalManager || GetRoleGroup(role) == RoleGroupRoot
}

func IsClassRole(role uint8) bool {
	return IsMember(role) || GetRoleGroup(role) == RoleGroupClassManager
}

func IsLoggedIn(role uint8) bool {
	return role != RoleGuest
}

func IsCertified(role uint8) bool {
	return role != RoleGuest && role != RoleRegularMember
}
